File: Boost_Echo_Client/src/StompFrame.cpp
//
// Created by yuval on 07/01/2020.
//

#include <vector>
#include <sstream>
#include <iostream>
#include "StompFrame.h"

StompFrame::StompFrame() : command(NONE), headers(std::map<std::string, std::string>()),
                           body(""), string_Enum_Convertor(std::map<std::string, Command>()),
                           enum_string_convertor(std::map<Command, std::string>()) {
    string_Enum_Convertor.insert(std::make_pair("CONNECTED", CONNECTED));
    string_Enum_Convertor.insert(std::make_pair("RECEIPT", RECEIPT));
    string_Enum_Convertor.insert(std::make_pair("MESSAGE", MESSAGE));
    string_Enum_Convertor.insert(std::make_pair("ERROR", ERROR));
    enum_string_convertor.insert(std::make_pair(CONNECT, "CONNECT"));
    enum_string_convertor.insert(std::make_pair(SUBSCRIBE, "SUBSCRIBE"));
    enum_string_convertor.insert(std::make_pair(UNSUBSCIRBE, "UNSUBSCRIBE"));
    enum_string_convertor.insert(std::make_pair(SEND, "SEND"));
    enum_string_convertor.insert(std::make_pair(DISCONNECT, "DISCONNECT"));
}

void StompFrame::parse(const std::string msg) {
    if (msg == "") {
//        std::cout <<"EMPTY MSG";
        return;
    } else {
        std::vector<std::string> lines;
        std::istringstream stream(msg);
        std::string line;
        while (std::getline(stream, line)) {
            lines.push_back(line);
        }
        command = string_Enum_Convertor[lines[0]];
        int i = 1;
        while ((unsigned) i < lines.size() && lines[i] != "") {
            int split = lines[i].find(':');
            headers.insert(std::make_pair(lines[i].substr(0, split), lines[i].substr(split + 1, lines[i].size())));
            i++;
        }
        if ((unsigned) i < lines.size() - 1)
            body = lines[i + 1];
        else
            body = "";
    }
}

std::string StompFrame::toString() {
    std::string output;
    output = enum_string_convertor[command] + "\n";
    for (auto p:headers) {
        output = output + p.first + ":" + p.second + "\n";

    }
    output = output + "" + "\n";
    output = output + body;
    return output;

}

void StompFrame::setCommand(Command command) {
    this->command = command;
}

void StompFrame::setBody(std::string body) {
    this->body = body;
}

void StompFrame::addHeader(std::string name, std::string value) {
    headers.insert(std::make_pair(name, value));
}

std::map<std::string, std::string> StompFrame::getHeaders() {
    return headers;
}

Command StompFrame::getCommand() {
    return command;
}

std::string StompFrame::getBody() {
    return body;
}

File: spl-net/src/main/java/bgu/spl/net/impl/stomp/StompFrame.java
package bgu.spl.net.impl.stomp;


import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

public class StompFrame {
    //fields
    public static enum Command {CONNECT, CONNECTED, MESSAGE, RECEIPT, ERROR, SEND, SUBSCRIBE, UNSUBSCRIBE, DISCONNECT,STOMP}
    private Command command;
    private LinkedHashMap<String,String> headers;
    private String body;

    public StompFrame(){
        headers = new LinkedHashMap<>();
        body = "";
    }

    public static StompFrame parse(String message){
        StompFrame f = new StompFrame();
        String[] mArr = message.split("\n");
        //set command
        f.command = Command.valueOf(mArr[0]);
        //set headers & body
        for (int i=1; i<mArr.length ; i++){
            if (mArr[i].length() == 0) {
                // "" signs end of headers
                if(mArr.length > i+1)
                    //there is another line which contains the StompBody
                    f.body = mArr[i+1];
                break;//there is nothing after body
            }
            else{
                //put header in headers map
                String[] header  = mArr[i].split(":");
                f.headers.put(header[0],header[1]);
            }
        }



        return f;
    }

    public String toString(){
        //Command
        String output = command.toString()+"\n";
        //Headers
        for (Map.Entry<String,String> entry : headers.entrySet()){
            String headerName = entry.getKey();
            String headerValue = entry.getValue();
            output = output + headerName+":"+headerValue+"\n";
        }
        //Body & endOfMesseage charcter
        output = output + "\n" + body;
        return output;
        }

    public void setCommand(String command) {
        this.command = Command.valueOf(command);
    }

    public void addHeader(String headerName, String headerValue) {
        headers.put(headerName,headerValue);
    }

    public void setBody(String body) {
        this.body = body;
    }

    public Command getCommand() {
        return command;
    }

    public HashMap<String, String> getHeaders() {
        return headers;
    }

    public String getBody() {
        return body;
    }
}
